### PR TITLE
More complete integer and floating point literals parsing.

### PR DESCRIPTION
**Pull request by prenaux**
I haven't made 100% certain to follow the whole grammar, its based on my memory, what is in the code I'm testing and a quick read of cppreference.com.

If there's interest I'd like to wrap this up with a proper regression test and check against the official grammar. Please see the TODO below for a few questions I have.

## TODO
- [ ] (Optional) Add an integration test. How are these ran on Linux/macOS? I only see a run-tests.bat, I can add the equivalent for bash (macOS & Linux) if there's interest to add those - and probably a GH action to actually run them after each push as-well while at it.
- [ ] (Optional) Check against the official standard grammar. Is there a link we can refer to? Is there an official test suite?

## Test code

```
void literals_cpp1() {
  // integers
  123;
  123u;
  123ul;
  123ull;
  123l;
  123ll;
  123'456ll;
  123;
  123U;
  123UL;
  123ULL;
  123L;
  123LL;
  123'456LL;
  // float points
  123'456.0f;
  123'456.f;
  456.0;
  456.;
  1.0e10;
  1.0e+10;
  1.0e-10;
  1.0e-10f;
  1.e-10;
  1.e-10f;
  1e-10;
  1e-10f;
  1e-1'0;
  123'456.0F;
  123'456.F;
  456.0;
  456.;
  1.0E10;
  1.0E+10;
  1.0E-10;
  1.0E-10F;
  1.E-10;
  1.E-10F;
  1E-10;
  1E-10F;
  1E-1'0;
  // invalid
  // 456f
  // (void)1e-10.0f;
  // 1.0e-10.0;
  // 1.0e-10.0f;
  // 1e-10.0;
  // 1e-10.0f;
  // 1e - 10;
  // 1e - 10.0;
  // 1e - 10f;
  // 1e - 10.0f;
}

literals_cpp2: () -> auto = {
  // integers
  123;
  123u;
  123ul;
  123ull;
  123l;
  123ll;
  123'456ll;
  123;
  123U;
  123UL;
  123ULL;
  123L;
  123LL;
  123'456LL;
  // float points
  123'456.0f;
  123'456.f;
  456.0;
  456.;
  1.0e10;
  1.0e+10;
  1.0e-10;
  1.0e-10f;
  1.e-10;
  1.e-10f;
  1e-10;
  1e-10f;
  1e-1'0;
  123'456.0F;
  123'456.F;
  456.0;
  456.;
  1.0E10;
  1.0E+10;
  1.0E-10;
  1.0E-10F;
  1.E-10;
  1.E-10F;
  1E-10;
  1E-10F;
  1E-1'0;
  // invalid
  // 456f
  // (void)1e-10.0f;
  // 1.0e-10.0;
  // 1.0e-10.0f;
  // 1e-10.0;
  // 1e-10.0f;
  // 1e - 10;
  // 1e - 10.0;
  // 1e - 10f;
  // 1e - 10.0f;
}
```

### CODE REVIEW

File: source/lex.h [hunk 1]
```diff
@@ -100,11 +100,11 @@ enum class lexeme : std::int8_t {
...
-    break;case lexeme::FloatLiteral:        
-          case lexeme::BinaryLiteral:       
-          case lexeme::DecimalLiteral:      
-          case lexeme::HexadecimalLiteral:  
-          case lexeme::StringLiteral:       
+    break;case lexeme::FloatLiteral:
+          case lexeme::BinaryLiteral:
+          case lexeme::DecimalLiteral:
+          case lexeme::HexadecimalLiteral:
+          case lexeme::StringLiteral:
```
Comment (prenaux) [inline source/lex.h:107]: My editor trims white spaces at the end of lines automatically, I can get rid of those diffs if they are an issue. I'm usually not a fan of "whitespace" only diffs...

File: source/lex.h [hunk 2]
```diff
@@ -825,29 +825,61 @@ auto lex_line(
...
 
+                //G
                 //G decimal-literal:
-                //G     digit { ' | digit }*
+                //G     digit { ' | digit }* [uU][lL][lL]
                 //G
                 //G floating-point-literal:
-                //G     digit { ' | digit }* . digit { ' | digit }*
-                //GTODO full grammar
+                //G     digit { ' | digit }* (. digit { ' | digit }*)? ([eE][-+]?digit { ' | digit }*) [fFlL]
```
Comment (prenaux) [inline source/lex.h:833]: I use a regex-like syntax for the grammar definition here, a bit sloppy, if there's a reference on the notation you prefer to use I'm happy to update it.

**Comment by JohelEGP** [comment]
GCC and Libstdc++ 13 implement [P1467R9](https://wg21.link/P1467R9), so tests for those to see what happens.

**Comment by jarzec** [comment]
In terms of running the tests on Linux/MacOS I had a [PR](https://github.com/hsutter/cppfront/pull/151). It included a simple Github action which was possibly the reason why it was closed as premature. I am still using an updated version of that bash script myself. I can push the updated version to a new branch later today and maybe PR it again without the Github action.

**Comment by filipsajdak** [comment]
I have run your code on mac, and it passes all regression tests.

**Comment by prenaux** [comment]
@jarzec the bash script in your PR looks good - I'd be glad to look at the updated version if you have one.

I'm ok with keeping those only in my fork. I can understand not wanting to deal with it in this repo since its extra work to maintain, and it can be quite a time sink...

**Comment by prenaux** [comment]
Rebased on main.

**Comment by hsutter** [comment]
Thanks! Looks reasonable, but if you could please check the grammar in http://eel.is/c++draft/ that would be great.

**Pull request merged**